includes/reader-revenue/woocommerce/class-woocommerce-products.php

```diff
@@ -0,0 +1,190 @@
+<?php
+/**
+ * Extensions for WooCommerce products.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+use WP_Error;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Connection with WooCommerce's features.
+ */
+class WooCommerce_Products {
+	/**
+	 * Initialize.
+	 *
+	 * @codeCoverageIgnore
+	 */
+	public static function init() {
+		\add_filter( 'product_type_options', [ __CLASS__, 'show_custom_product_options' ] );
+		\add_action( 'woocommerce_variation_options', [ __CLASS__, 'show_custom_variation_options' ], 10, 3 );
+		\add_action( 'woocommerce_process_product_meta', [ __CLASS__, 'save_custom_product_options' ] );
+		\add_action( 'woocommerce_admin_process_variation_object', [ __CLASS__, 'save_custom_variation_options' ], 30, 2 );
+		\add_filter( 'woocommerce_order_item_needs_processing', [ __CLASS__, 'require_order_processing' ], 10, 2 );
+	}
+
+	/**
+	 * Get custom product options. Product option values are always booleans but represented by string.
+	 * See: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-formatting-functions.php#L37
+	 *
+	 * @return array Keyed array of custom product options.
+	 */
+	public static function get_custom_options() {
+		return [
+			'newspack_autocomplete_orders' => [
+				'id'            => '_newspack_autocomplete_orders',
+				'wrapper_class' => 'show_if_simple',
+				'label'         => __( 'Auto-complete orders', 'newspack-plugin' ),
+				'description'   => __( 'Allow orders containing this product to automatically complete upon successful payment.', 'newspack-plugin' ),
+				'default'       => 'yes',
+			],
+		];
+	}
+
+	/**
+	 * Get the value of a custom product option, taking defaults into account.
+	 *
+	 * @param \WC_Product|int $product The product object or ID.
+	 * @param string          $option_name The name of the option.
+	 *
+	 * @return bool|null The value of the option as converted by wc_string_to_bool,
+	 *                   or null if the product or option isn't valid.
+	 */
+	public static function get_custom_option_value( $product, $option_name ) {
+		$custom_options = self::get_custom_options();
+		if ( ! isset( $custom_options[ $option_name ] ) ) {
+			return null;
+		}
+
+		$product = is_a( $product, 'WC_Product' ) ? $product : \wc_get_product( $product );
+		if ( ! $product ) {
+			return null;
+		}
+
+		$custom_option = $custom_options[ $option_name ];
+		$meta_key      = $custom_option['id'];
+		$option_value  = $product->meta_exists( $meta_key ) ? $product->get_meta( $meta_key ) : $custom_option['default'];
```

Comment:
Shouldn't the true-by-default be conditional on the product being a Newspack Donation product?

Reply:
The impetus for this feature was that several publishers implemented custom subscription-based products but didn't make them virtual + downloadable, and therefore weren't syncing contact data upon purchase (because the orders were getting stuck in "processing" status). So we want the defaults to apply across the board to cover this case.